pkg/context/context.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package context
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// A Context carries a deadline, a cancellation signal, and other values across
+// API boundaries.
+//
+// Context's methods may be called by multiple goroutines simultaneously.
+type Context interface {
```

Comment:
cc @munnerz I think you did something similiar for cert-manager?

Reply:
Yes, although we specifically avoided doing what's been done here and adding a new Context type, instead using accessors that accept regular contexts in order to 'read' or 'write' the appropriate logger etc: https://github.com/jetstack/cert-manager/blob/master/pkg/logs/logs.go#L132

I've not had a chance to properly review this PR, but I'd be strongly -1 to creating a new Context type as it'll make interoperability with other packages harder, as the regular 'ctx' being passed into a caller would not be able to just hand this off to a controller-runtime function.

By keeping the logger passing and the context handling isolated from each other (by just making use of context keys to store the logger), it's easy to integrate existing code with cert-manager packages. If that context does not contain a Logger, the 'default' is instead used (either a NullLogger or the 'global' default one).

I'll try and take a proper look at this PR later this week, but just a few thoughts based on our experience 😄

Reply:
IMO a context isn't being used for what it is designed for if we are embedding it into another interface. I'm no expert but it feels like a bit of a code smell to me.

If this is truly the approach we want, in order to ensure that the Log is always explicitly provided by the caller (so no defaulting behaviour happens), it seems more explicit to just pass in the Logger as an argument where required.

In practice though I would lean towards making use of context keys 🙂 then at the start of methods that need it, you'd call something like `log := logf.FromContext(ctx)`.

The only frustrating part here (which is also true for this design too) is the need to re-wrap loggers with contexts if a 'child' logger is made. We have a `NewContext` function to make that easier: https://github.com/jetstack/cert-manager/blob/master/pkg/logs/logs.go#L144

Reply:
Thanks for the context around context @munnerz 😂, we'll have to think about it a little more and so far we're experimenting and see how we'd like to proceed.

> as it'll make interoperability with other packages harder

Can you provide an example for this? 

@luxas @alvaroaleman what do you all think?

Reply:
Well I pinged james after I randomly had a discussion about this with someone yesterday who said mostly the same, the idea of the context is to hold all context and if we now build our own context, we basically break eco-system compatibility because everything else uses the `context.Context` (and that can not be used to retrieve the logger anymore). So I'd vote for not having our own context as well.

Reply:
I'm fine moving the logger somewhere else, @alvaroaleman and I discussed to potentially move it under the reconcile.Request struct if that makes sense and use stdlib's context.Context. The other option (shown here), if we want to go down the path of having our own context library that's a drop-in replacement for the stdlib's one, but be explicit that this should only be used for the current request.

Personally, I'd like to be as explicit as possible and probably not have the drop-in package. If the logger is bundled with the Request, it's probably going to be very clear that that logger is associated with the object that came in. 

Let's hear from more folks though, I'd say I feel a little meh about both approaches 😂

Reply:
+1 to using a func with `context.Context` as a param to access extended functionality `to.Logger(ctx).Info()` where `ctx` is a `context.Context`.

You lose a little bit of convenience by not being able to say `ctx.Logger()`, but having a vanilla `context.Context` will be natural to anyone familiar to the stdlib.

Reply:
+1 to using the vanilla `context.Context`. 

The grpc-middleware project has a nice example on how it can be done:
https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/logging/zap/ctxzap/context.go#L35